### ISSUE TITLE
Allow running behind TLS terminating reverse proxy

### ISSUE THREAD

**Issue by cnschn** [greenpau/caddy-auth-portal#62]
I'm deploying caddy on Kubernetes behind an ingress controller, i.e. Nginx, which handles TLS termination. In principle, `caddy-auth-portal` works fine in this setup, but the OAuth2 `redirect_url` is built using the wrong scheme, because `GetCurrentBaseURL` just checks the `TLS` field on the `Response`, and returns `http://domain`  since there is no TLS between Nginx and caddy.

Is there some way to force the `redirect_url` to use the https scheme? Either by explicitly overwriting the whole redirect, or by using something like the [`SECURE_PROXY_SSL_HEADER`](https://docs.djangoproject.com/en/3.1/ref/settings/#secure-proxy-ssl-header) in Django, which uses a header like `X-Forwarder-Proto` to decide if the request was https originally?

**Comment by greenpau** [comment]
@cnschn , will work through this :-)

1. First, what is your current config?
2. What  is the current `redirect_url` value?
3. What is the desired value (or logic)?

The diagram would help a lot! You could start one here: https://www.diagrams.net/

If whatever you are looking for does not exist, I will create it (as long as I can understand the ask).

**Comment by cnschn** [comment]
Great, thank you @greenpau. Here's the diagram:
![caddy-auth-portal-in-k8s](https://user-images.githubusercontent.com/1759343/99485086-09083a00-2962-11eb-8d98-e0b5163b23ce.png)

Here's the Caddyfile:

```
{
  https_port 8443
  http_port 8000
}

:8000 {
  route {
    jwt {
      primary yes
      trusted_tokens {
        static_secret {
          token_name access_token
          token_secret {$TOKEN_SECRET}
        }
      }
      auth_url /auth
    }

    respond "Here there be content!"
  }

  route /auth* {
    auth_portal {
      path /auth
      ui {
        theme basic
      }
      jwt {
        token_name access_token
        token_secret {$TOKEN_SECRET}
      }
      github_oauth2_backend {
        method oauth2
        realm github
        provider github
        client_id {$CLIENT_ID}
        client_secret {$CLIENT_SECRET}
        scopes user
      }
    }
  }
}
```

(I have to change `http(s)_port` because the cluster I'm using forces non-root users, so no binding to any ports < 1000)

This makes caddy-auth-portal build the `redirect_url` `http://my-application.apps.example.com/auth/oauth2/github/authorization-code-callback`, the correct one being `https://my-application.apps.example.com/auth/oauth2/github/authorization-code-callback`. This is understandable, while caddy-auth-portal gets the "outside" host from the http request, it has no way of knowing if the original scheme was `https`, only that the connection from the ingress controller to the caddy container was made using `http`.

Some ways out I could think of:

1. provide a configuration option to override the complete `redirect_url`, or probably rather something like a `redirect_base_url`, withouth the `/auth/oauth2/...`
2. provide a configuration option just to override the scheme
3. implement logic that takes the scheme from something like the `X-Forwarded-Proto` header

Option 2 would be the easiest for me right now, but I'm not sure what would be most useful in general.

**Comment by greenpau** [comment]
@cnschn , I can get you options 1 and 2 tomorrow. It is easy and reasonable.

As per option 3, please enable https://github.com/greenpau/caddy-trace plugin and get me the trace where I can see received headers. If you get me that, then I will implement it too.

Like I said it is pretty easy doing it. So expect it tomorrow EOD EST.

Another question, do you use “caddy-auth-jwt”?

**Comment by cnschn** [comment]
Awesome, sounds great!

Here's the `"headers"` key from the trace (really useful plugin as well, by the way!)
```
"headers": {
    "Accept": "text/html,application/xhtml+xml,application/xml;q=0.9,image/avif,image/webp,image/apng,*/*;q=0.8,application/signed-exchange;v=b3;q=0.9",
    "Accept-Encoding": "gzip, deflate, br",
    "Accept-Language": "en-US,en;q=0.9,de;q=0.8",
    "Sec-Fetch-Dest": "document",
    "Sec-Fetch-Mode": "navigate",
    "Sec-Fetch-Site": "none",
    "Sec-Fetch-User": "?1",
    "Upgrade-Insecure-Requests": "1",
    "User-Agent": "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/86.0.4240.198 Safari/537.36",
    "X-Forwarded-For": "2.243.31.240",
    "X-Forwarded-Host": "docs.ttt.000.westeurope.containers.sbp.eyclienthub.com",
    "X-Forwarded-Port": "443",
    "X-Forwarded-Proto": "https",
    "X-Real-Ip": "2.243.31.240",
    "X-Request-Id": "ac8b0c37428de700f8bf76ee82b8e813",
    "X-Scheme": "https"
  },

```

And yes, I'm also using caddy-auth-jwt, in combination with auth-portal this is a really convenient setup.

**Comment by greenpau** [comment]
@cnschn , I am going to add the following headers for the calculation of redirect URL. This way there is no need to add additional configuration.

```
"X-Forwarded-Host": "my-application.apps.example.com",
"X-Forwarded-Port": "443",
"X-Forwarded-Proto": "https",
```

**Comment by greenpau** [comment]
@cnschn , I implemented the necessary changes [here](https://github.com/greenpau/caddy-auth-jwt/commit/4c12537a4abd8581370095031d09457ac423d2ee).

You need to use the following versions to benefit from the feature:

```
caddy-auth-jwt v1.1.10
caddy-auth-portal v1.2.20
```

Please help promoting the project by posting about it on the social media platforms you use! 👍

**Comment by cnschn** [comment]
> @cnschn , I implemented the necessary changes [here](https://github.com/greenpau/caddy-auth-jwt/commit/4c12537a4abd8581370095031d09457ac423d2ee).

@greenpau I'm sorry for not properly checking this earlier, but I fear we have been speaking about different redirect URLs... While the new logic in `caddy-auth-jwt` is definitely good to have, the primary issue I am having is with the `redirect_uri` URL parameter sent to the OAuth backend, i.e. here:

https://github.com/greenpau/caddy-auth-portal/blob/9e61321d98a09009366c1a0cd2e7deb327ab9705/pkg/backends/oauth2/backend.go#L479-L489

It is built here

https://github.com/greenpau/caddy-auth-portal/blob/9e61321d98a09009366c1a0cd2e7deb327ab9705/pkg/backends/oauth2/backend.go#L263

using

https://github.com/greenpau/caddy-auth-portal/blob/9e61321d98a09009366c1a0cd2e7deb327ab9705/pkg/utils/redirect.go#L32-L38

Which doesn't give the correct result, as explained initally.

Also, one more thing, I'm not sure if it's a good idea to explicitly add the port to the url as you did in `caddy-auth-jwt`, at least not for `80`/`443`, as the OAuth backends typically only do string comparison, so `https://example.com/whatever != https://example.com:443/whatever`.

**Comment by greenpau** [comment]
>  I am having is with the redirect_uri URL parameter sent to the OAuth backend, i.e. here:

cnschn , yes, that was a misunderstanding.

> Also, one more thing, I'm not sure if it's a good idea to explicitly add the port to the url as you did in caddy-auth-jwt, at least not for 80/443, as the OAuth backends typically only do string comparison, 

I think this should be be a caddy directive, i.e. optional.

Are you planning a PR for this?

**Comment by cnschn** [comment]
> I think this should be be a caddy directive, i.e. optional.

I agree.

In this "proxied" situation, only the _scheme_ of the redirect URI is giving problems, everything else works fine. This makes sense, since the hostname and port are passed in the `Host` header.

So in the interest of not over-engineering this, I'd propose a simple boolean `redirect_uri_force_https` (not sure about the name) directive, which would override the scheme to https. I don't think there is a use case for the reverse, so a directive to arbitrarily _set_ the scheme seems weird to me. Unless you would prefer something more general?

> Are you planning a PR for this?

Sure, I can try to cook something up :)

**Comment by greenpau** [comment]
> So in the interest of not over-engineering this, I'd propose a simple boolean redirect_uri_force_https (not sure about the name) directive, which would override the scheme to https. I don't think there is a use case for the reverse, so a directive to arbitrarily set the scheme seems weird to me. Unless you would prefer something more general?

@cnschn , let me think through this. There has to be an easier way to address this.

**Comment by greenpau** [comment]
@cnschn , I added the fix [here](https://github.com/greenpau/caddy-auth-portal/commit/f374702031412c886bff0e273470e2528962734e).

It is available in [v1.2.23](https://github.com/greenpau/caddy-auth-portal/releases/tag/v1.2.23).

Please test and see if it works as expected.

**Comment by cnschn** [comment]
It works! At least for the azure backend, which is what I can test right now.

> Also, one more thing, I'm not sure if it's a good idea to explicitly add the port to the url as you did in `caddy-auth-jwt`, at least not for `80`/`443`, as the OAuth backends typically only do string comparison, so `https://example.com/whatever != https://example.com:443/whatever`.

I'm still concerned about this, though - while Azure surprisingly (to me) does not care about the `:443` substring, I'm not sure if all backends act the same way.

**Comment by greenpau** [comment]
Do you mean http redirect has port 443 in it? That’s easy to adjust.

**Comment by cnschn** [comment]
Exactly, I'd suggest not adding the Port to the string, unless it's something other than 80 or 443.

**Comment by greenpau** [comment]
> Exactly, I'd suggest not adding the Port to the string, unless it's something other than 80 or 443.

@cnschn , I added the fix in [v1.3.1](https://github.com/greenpau/caddy-auth-portal/releases/tag/v1.3.1). Please test.

**Comment by cnschn** [comment]
Works perfectly @greenpau, thanks again!
